modules/siddhi-extensions/markov-models/src/main/java/org/wso2/siddhi/extension/markov/MarkovChainTransitionProbabilitiesCalculator.java

```diff
@@ -0,0 +1,237 @@
+import java.io.*;
+import java.util.*;
+
+public class MarkovChainTransitionProbabilitiesCalculator implements Serializable {
+
+    private static final long serialVersionUID = -7632888859752532839L;
+    private static final Logger log = Logger.getLogger(MarkovChainTransitionProbabilitiesCalculator.class);
+    private long durationToKeep;
+    private long notificationsHoldLimit;
+    private double alertThresholdProbability;
+    private int eventCount;
+    private Map<String, String> idToLastStates = new HashMap<String, String>();
+    private Map<String, Long> idToLastStatesExpiryTime = new LinkedHashMap<String, Long>();
+    private MarkovMatrix markovMatrix = new MarkovMatrix();
+
+    /**
+     * Constructor for real time training method
+     *
+     * @param durationToKeep
+     * @param alertThresholdProbability
+     * @param notificationsHoldLimit
+     */
+    public MarkovChainTransitionProbabilitiesCalculator(long durationToKeep, double alertThresholdProbability,
+            long notificationsHoldLimit) {
+        this.durationToKeep = durationToKeep;
+        this.notificationsHoldLimit = notificationsHoldLimit;
+        this.alertThresholdProbability = alertThresholdProbability;
+    }
+
+    /**
+     * Constructor for predefined markov matrix method
+     *
+     * @param durationToKeep
+     * @param alertThresholdProbability
+     * @param markovMatrixStorageLocation
+     */
+    public MarkovChainTransitionProbabilitiesCalculator(long durationToKeep, double alertThresholdProbability,
+            String markovMatrixStorageLocation) {
+        this.durationToKeep = durationToKeep;
+        this.alertThresholdProbability = alertThresholdProbability;
+        populateMarkovMatrix(markovMatrixStorageLocation);
+    }
+
+    /**
+     * Method to process new events
+     *
+     * @param userId
+     * @param state
+     * @param trainingOption
+     * @return Array containing output results
+     */
+    protected Object[] processData(String userId, String state, boolean trainingOption) {
+
+        eventCount++;
+        String id = userId;
+        String currentState = state;
+        boolean currentTrainingOption = trainingOption;
+        Object[] markovChainAlert = null;
+
+        String lastState = updateStates(id, currentState);
+
+        if (eventCount > notificationsHoldLimit) {
+            markovChainAlert = getMarkovChainAlert(lastState, currentState);
+        }
+
+        if (currentTrainingOption && lastState != null) {
+            markovMatrix.updateStartStateCount(lastState, 1);
+            markovMatrix.updateTransitionCount(lastState, currentState, 1);
+        }
+
+        return markovChainAlert;
+    }
+
+    /**
+     * Method to updateTransitionCount States
+     */
+    private String updateStates(String id, String currentState) {
+
+        String lastState = idToLastStates.get(id);
+        // updateTransitionCount the last state to current state
+        idToLastStates.put(id, currentState);
```

Comment:
Thanks for the suggestions. Fixed with https://github.com/wso2/siddhi/pull/189/commits/690b60924a459f505cce35f57d2ce6b05e4980c2